src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -254,9 +281,29 @@ public JsonClassInfo ElementClassInfo
             return (TAttribute)propertyInfo?.GetCustomAttribute(typeof(TAttribute), inherit: false);
         }
 
+        public abstract Type GetConcreteType(Type type);
+
         public abstract Type GetDictionaryConcreteType();
 
-        public abstract Type GetConcreteType(Type type);
+        public void GetDictionaryKeyAndValue(ref WriteStackFrame writeStackFrame, out string key, out object value)
+        {
+            Debug.Assert(ClassType == ClassType.Dictionary ||
+                ClassType == ClassType.IDictionaryConstructible);
+
+            if (writeStackFrame.CollectionEnumerator is IDictionaryEnumerator iDictionaryEnumerator)
+            {
```

Comment:
Would it be beneficial to perform this check at the calling site of this method, since we don't need the generic support? That way we don't need to call into the `JsonPropertyInfo`.

On the other hand, would performing the generic check done in `DictionaryValuePropertyPolicy.GetDictionaryKeyAndValueFromGenericDictionary()` first (before checking for `IDictionaryEnumerator`) have a perf benefit (by avoiding boxing) for types whose `.GetEnumerator()` method implements both `IDictionaryEnumerator` and `IEnumerator<KeyValuePair<,>>` (e.g. `Dictionary<,>`)?

Reply:
It is called from more than one location, so keeping it in the helper method is a bit cleaner.

On the generic check, I think it is better to check for non-generic first in this case as the generic check will unnecessary create a "policy property" (which is expensive) in the non-generic cases.

Also, as a separate PR I'm working on avoid boxing in general for this case for perf...